crates/katana/executor/src/implementation/blockifier/state.rs

```diff
@@ -68,29 +66,29 @@ impl<'a> StateReader for StateProviderDb<'a> {
     }
 
     fn get_nonce_at(
-        &mut self,
+        &self,
         contract_address: katana_cairo::starknet_api::core::ContractAddress,
     ) -> StateResult<katana_cairo::starknet_api::core::Nonce> {
         self.0
             .nonce(utils::to_address(contract_address))
-            .map(|n| Nonce(to_stark_felt(n.unwrap_or_default())))
+            .map(|n| Nonce(n.unwrap_or_default()))
             .map_err(|e| StateError::StateReadError(e.to_string()))
     }
 
     fn get_storage_at(
-        &mut self,
+        &self,
         contract_address: katana_cairo::starknet_api::core::ContractAddress,
         key: katana_cairo::starknet_api::state::StorageKey,
-    ) -> StateResult<katana_cairo::starknet_api::hash::StarkFelt> {
+    ) -> StateResult<katana_cairo::starknet_api::hash::StarkHash> {
         self.0
-            .storage(utils::to_address(contract_address), to_felt(*key.0.key()))
-            .map(|v| to_stark_felt(v.unwrap_or_default()))
+            .storage(utils::to_address(contract_address), *key.0.key())
+            .map(|v| v.unwrap_or_default())
             .map_err(|e| StateError::StateReadError(e.to_string()))
     }
 }
 
 #[derive(Debug)]
-pub(super) struct CachedState<S: StateDb>(pub(super) Arc<RwLock<CachedStateInner<S>>>);
+pub(super) struct CachedState<S: StateDb>(pub(super) Arc<Mutex<CachedStateInner<S>>>);
```

Comment:
`Mutex` instead of `RwLock` because you found that it was less "read-heavy" as expected and much writes are expected too?
Or there's an other reason for this change?

Reply:
no, bcs that `RwLock` requires `T` to be both `Send` and `Sync` which due to the use of `RefCell` in `CachedState` (the `T` value), it is now `!Sync`. 

`Mutex` has more relaxed restriction requiring `T` to only be `Send.

Reply:
though i think we should not naively put a lock there (doesn't matter if its Mutex/RwLock). would still hurt performance but a Mutex would definitely hurt the most. i should consider something less restricting